components/engine_rocks/src/properties.rs

```diff
@@ -493,6 +494,120 @@ impl TablePropertiesCollector for MvccPropertiesCollector {
     }
 }
 
+/// Can only be used for write CF.
+pub struct RawMvccPropertiesCollector {
+    props: MvccProperties,
+    last_row: Vec<u8>,
+    num_errors: u64,
+    row_versions: u64,
+    cur_index_handle: IndexHandle,
+    row_index_handles: IndexHandles,
+}
+
+impl RawMvccPropertiesCollector {
+    fn new() -> RawMvccPropertiesCollector {
+        RawMvccPropertiesCollector {
+            props: MvccProperties::new(),
+            last_row: Vec::new(),
+            num_errors: 0,
+            row_versions: 0,
+            cur_index_handle: IndexHandle::default(),
+            row_index_handles: IndexHandles::new(),
+        }
+    }
+}
+
+impl TablePropertiesCollector for RawMvccPropertiesCollector {
+    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
+        // TsFilter filters sst based on max_ts and min_ts during iterating.
+        // To prevent seeing outdated (GC) records, we should consider
+        // RocksDB delete entry type.
+        if entry_type != DBEntryType::Put && entry_type != DBEntryType::Delete {
+            return;
+        }
+
+        let current_ts = ttl_current_ts();
+
+        let key_mode = ApiV2::parse_key_mode(keys::origin_key(key));
+        if key_mode != KeyMode::Raw {
+            return;
+        }
+
+        if !keys::validate_data_key(key) {
+            self.num_errors += 1;
+            return;
+        }
+
+        let (k, ts) = match ApiV2::split_ts(key) {
+            Ok((k, ts)) => (k, ts),
+            Err(_) => {
+                self.num_errors += 1;
+                return;
+            }
+        };
+
+        self.props.min_ts = cmp::min(self.props.min_ts, ts);
+        self.props.max_ts = cmp::max(self.props.max_ts, ts);
+        if entry_type == DBEntryType::Delete {
+            // Empty value for delete entry type, skip following properties.
+            return;
+        }
+
+        self.props.num_versions += 1;
+
+        if k != self.last_row.as_slice() {
+            self.props.num_rows += 1;
+            self.row_versions = 1;
+            self.last_row.clear();
+            self.last_row.extend(k);
+        } else {
+            self.row_versions += 1;
+        }
+        if self.row_versions > self.props.max_row_versions {
+            self.props.max_row_versions = self.row_versions;
+        }
+
+        let decode_raw_value = ApiV2::decode_raw_value(value);
+        if let Ok(raw_value) = decode_raw_value {
+            if raw_value.is_valid(current_ts) {
+                self.props.num_puts += 1;
+                if !raw_value.is_delete {
+                    let expire_ts = raw_value.expire_ts.unwrap();
```

Comment:
Suggest not to assume that a not deleted entry must have `expire_ts`. We may add more fields in the future.
Also, `expire_ts == 0` means never expired, should be ignored.

Reply:
I modified the logic, If expire_ts is None, it will not affect min_ttl_ts